### PR TITLE
tools, media, readme: add DM shard merge

### PR DESCRIPTION
**Pull request by lilin90**
This PR adds the DM shard merge file, updates the TOC, adds the images that are ignored in #886.

**Note:** Since the file in #876 (created by @yikeke) has links that point to the file added in this PR, #876 can only be merged after this PR is merged.

@GregoryIan @csuzhangxc @CaitinChen @yikeke PTAL

Related Issue: https://github.com/pingcap/docs/issues/874

Related PRs: 
- https://github.com/pingcap/docs/pull/876
- https://github.com/pingcap/docs/pull/886
- https://github.com/pingcap/tidb-tools/pull/161

### CODE REVIEW

**Comment by yikeke** [comment]
LGTM

File: tools/dm/shard-merge.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    - For example, if you add `column A` to `table_1` before you add `column B`, then you cannot add `column B` to `table_2` before you add `column A`. Executing the DDL statements in a different order is not supported.
+- If multiple sharding groups exist in a task, you cannot start to execute the DDL statements in other sharding groups until the DDL statements in one sharding group has been replicated successfully.
+    - For each sharding group, it is recommended to use one independent task to perform the replication.
+- In a sharding group, the corresponding DDL statements should be executed in all upstream sharded tables.
+    - For example, if DDL statements are not executed on one or more upstream sharded tables corresponding to `DM-worker-2`, then other DM-workers that have executed the DDL statements will pause their synchronization task and wait for `DM-worker-2` to receive the upstream DDL statements.
+- In a sharding group, the next DDL statement cannot be executed unless the current DDL operation is completely finished.
```
Comment (GregoryIan) [inline tools/dm/shard-merge.md]: merge it with `In a logical **sharding group** (composed of all sharded tables that need to be merged and replicated into one same downstream table), the same DDL statements must be executed in the same order in all upstream sharded tables (the schema name and the table name can be different)` at L17?

File: tools/dm/shard-merge.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+- If you need to change the [table routing](../../tools/dm/data-synchronization-features.md#table-routing) rule, you have to wait for the replication of all sharding DDL statements to complete.
+    - During the replication of sharding DDL statements, an error is reported if you use `dmctl` to change `router-rules`.
+- If you need to `CREATE` a new table to a sharding group where DDL statements are being executed, you have to make sure that the table schema is the same as the newly modified table schema.
+    - For example, the original `table_1` and `table_2` has two columns (a, b) initially, and has three columns (a, b, c) after the sharding DDL operation, so after the replication the newly created table should also have three columns (a, b, c).
+- Because the DM-worker that has received the DDL statements will pause the task to wait for other DM-workers to receive their DDL statements, the data replication will be increased.
```
Comment (GregoryIan) [inline tools/dm/shard-merge.md]: what's meaning of `the data replication will be increased.`?

File: tools/dm/shard-merge.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+- The sharding group replication task does not support `DROP DATABASE/DROP TABLE`.
+    - The Syncer unit in DM-worker automatically ignores the `DROP DATABASE/DROP TABLE` statement of upstream sharded tables.
+- The sharding group replication task supports `RENAME TABLE`, but with the following limitations:
+    - A table can only be renamed to a new name that is not used by any other table.
+    - A single `RENAME TABLE` statement can only involve a single `RENAME` operation. (Online DDL is supported in another solution)
+- The table schema of each sharded table must be the same at the starting point of the incremental replication task, so as to make sure the DML statements of different sharded tables can be replicated into the downstream with a definite table schema, and the subsequent sharding DDL statements can be correctly matched and replicated.
+- If you need to change the [table routing](../../tools/dm/data-synchronization-features.md#table-routing) rule, you have to wait for the replication of all sharding DDL statements to complete.
+    - During the replication of sharding DDL statements, an error is reported if you use `dmctl` to change `router-rules`.
```
Comment (GregoryIan) [inline tools/dm/shard-merge.md]: do we have a test case for it? @amyangfei @csuzhangxc

File: tools/dm/shard-merge.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- If multiple sharding groups exist in a task, you cannot start to execute the DDL statements in other sharding groups until the DDL statements in one sharding group has been replicated successfully.
+    - For each sharding group, it is recommended to use one independent task to perform the replication.
+- In a sharding group, the corresponding DDL statements should be executed in all upstream sharded tables.
+    - For example, if DDL statements are not executed on one or more upstream sharded tables corresponding to `DM-worker-2`, then other DM-workers that have executed the DDL statements will pause their synchronization task and wait for `DM-worker-2` to receive the upstream DDL statements.
+- In a sharding group, the next DDL statement cannot be executed unless the current DDL operation is completely finished.
+- The sharding group replication task does not support `DROP DATABASE/DROP TABLE`.
+    - The Syncer unit in DM-worker automatically ignores the `DROP DATABASE/DROP TABLE` statement of upstream sharded tables.
+- The sharding group replication task supports `RENAME TABLE`, but with the following limitations:
```
Comment (GregoryIan) [inline tools/dm/shard-merge.md]: I think we should add `(Online DDL is supported in another solution)` here

File: tools/dm/shard-merge.md (hunk 3 above)
Comment (csuzhangxc) [reply tools/dm/shard-merge.md]: I think no test case exists, and we only add a check in DM.

**Comment by GregoryIan** [comment]
Rest LGTM

File: tools/dm/shard-merge.md (hunk 3 above)
Comment (GregoryIan) [reply tools/dm/shard-merge.md]: we should make sure it works

File: README.md [hunk 5]
```diff
@@ -123,29 +123,29 @@
...
+      - [Binlog Event Filtering](tools/dm/data-synchronization-features.md#binlog-event-filtering)
+      - [Column Mapping](tools/dm/data-synchronization-features.md#column-mapping)
+      - [Synchronization Delay Monitoring](tools/dm/data-synchronization-features.md#synchronization-delay-monitoring)
       + Sharding Support
-        - [Introduction](/tools/dm/sharding-solution.md)
-        - [Restrictions](/tools/dm/sharding-solution.md#sharding-ddl-usage-restrictions)
-        - [Troubleshoot](/tools/dm/troubleshooting-sharding-ddl-locks.md)
+        - [Introduction](tools/dm/shard-merge.md)
+        - [Restrictions](tools/dm/shard-merge.md#restrictions)
+        - [Handle Sharding DDL Locks manually](tools/dm/manually-handling-sharding-ddl-locks.md)
```
Comment (csuzhangxc) [inline README.md]: link to file added in #876. so we need to merge #876 as soon as this PR merged?

File: tools/dm/shard-merge.md (hunk 2 above)
Comment (csuzhangxc) [reply tools/dm/shard-merge.md]: maybe *the delay of data replication*?

File: tools/dm/shard-merge.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+- After being converted by the [table router](../../tools/dm/data-synchronization-features.md#table-routing), the DDL statement of the upstream sharded tables must be consistent with the DDL statement to be executed in the downstream. Therefore, this DDL statement only needs to be executed once by the DDL owner and all other DM-workers can ignore this DDL statement.
+
+In the above example, only one sharded table needs to be merged in the upstream MySQL instance corresponding to each DM-worker. But in actual scenarios, there might be multiple sharded tables in multiple sharded schemas to be merged in one MySQL instance. And when this happens, it becomes more complex to coordinate the sharding DDL replication.
+
+Assume that there are two sharded tables, namely `table_1` and `table_2`, to be merged in one MySQL instance:
+
+![shard-ddl-example-2](../../media/shard-ddl-example-2.png)
+
+Because data comes from the same MySQL instance, all the data is obtained from the same binlog flow. In this case, the time sequence is as follows:
```
Comment (csuzhangxc) [inline tools/dm/shard-merge.md]: how about *binlog stream*?

File: README.md (hunk 5 above)
Comment (lilin90) [reply README.md]: Yes.

**Comment by lilin90** [comment]
@GregoryIan @csuzhangxc Thanks! PTAL again

File: tools/dm/shard-merge.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+In DM, the simplified replication process of sharding DDL statements within the DM worker is as follows:
+
+1. When receiving the DDL statement of `table_1` at `t1`, the DM-worker records the DDL information and the current position of the binlog.
+2. DM-worker continues parsing the binlog between `t2` and `t3`.
+3. DM-worker ignores the DML statement with the `schema V2` schema that belongs to `table_1`, and replicates the DML statement with the `schema V1` schema that belongs to `table_2` to the downstream.
+4. When receiving the DDL statement of `table_2` at `t3`, the DM-worker records the DDL information and the current position of the binlog.
+5. Based on the information of the replication task configuration and the upstream schemas and tables, the DM-worker confirms that the DDL statements of all sharded tables in the MySQL instance have been received. If all DDL statements have been received, DM-worker replicates them to the downstream to modify the downstream table schema.
+6. DM-worker sets the starting point of parsing the new binlog flow to be the position saved at step #1.
```
Comment (csuzhangxc) [inline tools/dm/shard-merge.md]:
```suggestion
6. DM-worker sets the starting point of parsing the new binlog stream to be the position saved at step #1.
```

**Comment by csuzhangxc** [comment]
LGTM

**Pull request merged**